sn_node_manager/src/daemon_control.rs

```diff
@@ -0,0 +1,237 @@
+        program: daemon_path,
+        args: vec![
+            OsString::from("--port"),
+            OsString::from(port.to_string()),
+            OsString::from("--address"),
+            OsString::from(address.to_string()),
+        ],
+        contents: None,
+        username: None,
+        working_directory: None,
+        environment: None,
+    };
+    service_control.install(install_ctx)?;
+    service_control.start(&service_name.to_string())?;
+
+    Ok(())
+}
+
+pub async fn restart_node_service(
+    node_registry: &mut NodeRegistry,
+    peer_id: PeerId,
+    retain_peer_id: bool,
+    rpc_client: &dyn RpcActions,
+    service_control: &dyn ServiceControl,
+) -> Result<()> {
+    let nodes_len = node_registry.nodes.len();
+    let current_node = node_registry
+        .nodes
+        .iter_mut()
+        .find(|node| node.peer_id.is_some_and(|id| id == peer_id))
+        .ok_or_eyre(format!("Could not find the provided PeerId: {peer_id:?}"))?;
+    node_control::stop(current_node, service_control)
+        .await
+        .map_err(|err| {
+            eyre!(
+                "Error while stopping node {:?} with: {err:?}",
+                current_node.service_name
+            )
+        })?;
+
+    if retain_peer_id {
+        // reuse the same port and root dir to retain peer id.
+        service_control
+            .uninstall(&current_node.service_name.clone())
+            .map_err(|err| {
+                eyre!(
+                    "Error while uninstalling node {:?} with: {err:?}",
+                    current_node.service_name
+                )
+            })?;
+        let install_ctx = node_control::InstallNodeServiceCtxBuilder {
+            local: current_node.local,
+            data_dir_path: current_node.data_dir_path.clone(),
+            genesis: current_node.genesis,
+            name: current_node.service_name.clone(),
+            node_port: current_node.get_safenode_port(),
+            bootstrap_peers: node_registry.bootstrap_peers.clone(),
+            rpc_socket_addr: current_node.rpc_socket_addr,
+            log_dir_path: current_node.log_dir_path.clone(),
+            safenode_path: current_node.safenode_path.clone(),
+            service_user: current_node.user.clone(),
+            env_variables: node_registry.environment_variables.clone(),
+        }
+        .execute()?;
+        service_control.install(install_ctx).map_err(|err| {
+            eyre!(
+                "Error while installing node {:?} with: {err:?}",
+                current_node.service_name
+            )
+        })?;
+        node_control::start(
+            current_node,
+            service_control,
+            rpc_client,
+            VerbosityLevel::Normal,
+        )
+        .await
+        .map_err(|err| {
+            eyre!(
+                "Error while starting node {:?} with: {err:?}",
+                current_node.service_name
+            )
+        })?;
+    } else {
+        // else start a new node instance.
```

Comment:
You don't actually appear to be restarting a node here: you're creating a new one. Also, the previous node is not being killed, as far as I can see. In my opinion, it would make more sense to have `remove` and `add` commands that correspond to the same commands on the manager, and you would issue those commands in the scenario where you want a new node.

Reply:
Right yeah. That makes more sense. Will modify it.